pandas/core/dtypes/concat.py

```diff
@@ -5,27 +5,90 @@
 
 import numpy as np
 
+from pandas._libs import lib
 from pandas._typing import ArrayLike, DtypeObj
 
 from pandas.core.dtypes.cast import find_common_type
 from pandas.core.dtypes.common import (
+    is_bool_dtype,
     is_categorical_dtype,
     is_dtype_equal,
     is_extension_array_dtype,
+    is_integer_dtype,
     is_sparse,
 )
 from pandas.core.dtypes.generic import ABCCategoricalIndex, ABCSeries
+from pandas.core.dtypes.missing import na_value_for_dtype
 
 from pandas.core.arrays import ExtensionArray
 from pandas.core.arrays.sparse import SparseArray
 from pandas.core.construction import array, ensure_wrapped_if_datetimelike
 
 
+class NullArrayProxy:
+    """
+    Proxy object for an all-NA array.
+
+    Only stores the length of the array, and not the dtype. The dtype
+    will only be known when actually concatenating (after determining the
+    common dtype, for which this proxy is ignored).
+    Using this object avoids that the internals/concat.py needs to determine
+    the proper dtype and array type.
+    """
+
+    ndim = 1
+
+    def __init__(self, n: int):
+        self.n = n
+
+    @property
+    def shape(self):
+        return (self.n,)
+
+    def to_array(self, dtype: DtypeObj, fill_value=lib.no_default) -> ArrayLike:
+        """
+        Helper function to create the actual all-NA array from the NullArrayProxy
+        object.
+
+        Parameters
+        ----------
+        arr : NullArrayProxy
+        dtype : the dtype for the resulting array
+        fill_value : scalar NA-like value
+            By default uses the ExtensionDtype's na_value or np.nan. For numpy
+            arrays, this can be overridden to be something else (eg None).
+
+        Returns
+        -------
+        np.ndarray or ExtensionArray
+        """
+        if is_extension_array_dtype(dtype):
+            empty = dtype.construct_array_type()._from_sequence([], dtype=dtype)
+            indexer = -np.ones(self.n, dtype=np.intp)
+            return empty.take(indexer, allow_fill=True)
+        else:
+            # when introducing missing values, int becomes float, bool becomes object
+            if is_integer_dtype(dtype):
+                dtype = np.dtype("float64")
+            elif is_bool_dtype(dtype):
+                dtype = np.dtype(object)
```

Comment:
L71-L74 are basically `maybe_promote(dtype)` (in fact im thinking of implementing a much smaller `ensure_dtype_can_hold_na` that would be almost exclusively this)

Reply:
I think in the meantime you added this? (will use it here then)